ceph/ceph.py

```diff
@@ -1391,8 +1394,50 @@ def generate_id_rsa(self):
             cmd="test -f ~/.ssh/id_rsa.pub && rm -f ~/.ssh/id*", check_ec=False
         )
         self.exec_command(cmd="ssh-keygen -b 2048 -f ~/.ssh/id_rsa -t rsa -q -N ''")
-        out1, _ = self.exec_command(cmd="cat ~/.ssh/id_rsa.pub")
-        self.id_rsa_pub = out1.read().decode()
+        self.id_rsa_pub, _ = self.exec_command(cmd="cat ~/.ssh/id_rsa.pub")
+
+    def long_running(self, **kw):
+        """Method to execute long running command.
+
+        Args:
+            **kw: execute command configuration
+
+        Returns:
+            read, ec
+        """
+        ssh = self.rssh if kw.get("sudo") else self.ssh
+        logger.info("long running command --")
+        channel = ssh().get_transport().open_session()
+        channel.exec_command(kw["cmd"])
+        read = ""
+        end_time = datetime.datetime.now() + timedelta(seconds=3600)
+
+        while end_time > datetime.datetime.now():
+            try:
+                if channel.exit_status_ready():
+                    ec = channel.recv_exit_status()
+                    break
+
+                rl, wl, xl = select([channel], [], [channel], 4200)
+
+                if len(rl) > 0:
+                    data = channel.recv(1024)
+                    data = data.decode(errors="ignore")
+                    logger.info(data)
+                    read += data
+
+                if len(xl) > 0:
+                    data = channel.recv(1024)
+                    data = data.decode(errors="ignore")
+                    logger.error(data)
+                    read += data
+            except socket.timeout as sock_err:
+                logger.error("socket.timeout doesn't give an error message")
+                raise SocketTimeoutException(sock_err)
+        else:
+            raise SocketTimeoutException("Timeout")
+        logger.info(f"start-time: {datetime.datetime.now()}")
+        return read, ec
```

Comment:
Can we avoid sending the data held in `read`? I believe this can turn to be large depending on how long the command is running and the level of verbosity.